WordPress/gutenberg
# Update to allow alternative Blocks to handle Grouping interactions

This PR will allow alternative Blocks to be used to handle Grouping interactions. Currently, you can only use `core/group` but this PR will make it possible to register any valid Block as the Block to use when Grouping.

Previously the method that determined whether a given Block was the block to be used for “Grouping” interactions was hardcoded as “core/group. This PR builds upon https://github.com/WordPress/gutenberg/pull/15774 to allow alternative Blocks to be used to handle the Grouping interactions.

This should address some of the concerns raised in https://github.com/WordPress/gutenberg/pull/14908#issuecomment-501534012, namely:

- [x] Avoid hardcoding refs to `core/group` within `packages/blocks/src/api/factory.js`
- [x] Allow other Blocks to be registered to handle Grouping if required

## Screenshots

Below shows me registering the `test/alternative-group-block` Block from a custom e2e testing Plugin as the default Grouping Block. I'm then using it within the Grouping interactions.

![Screen Capture on 2019-07-03 at 13-19-25](https://user-images.githubusercontent.com/444434/60590978-44b65a00-9d95-11e9-9f17-daa778927e0f.gif)


## How has this been tested?

* Unit tests
* e2e tests
* Manual testing using the method outlined below.

## How to test

* Open `wp-admin` in the Gutenberg docker env
* Activate the `Gutenberg Test Custom Grouping Block` Plugin (**important**: be sure to _deactivate_ this after you have finished testing!)
* Create a new Page and add some Blocks to be Grouped.
* In the browser console 
  * type `window.wp.blocks.getGroupingBlockName()` to get the current Grouping Block name - it should be `core/group` (the default).
  * type `wp.blocks.setGroupingBlockName( 'test/alternative-group-block' );` - it should register that block (provided by the Plugin activated above) as the default for Grouping
* using the Editor UI
  * "Group" the Blocks using the "Group" item from the dropdown menu
  * Ungroup the Blocks again
  * Group the Blocks using the "Transform" control by selecting the "Alternative Group Block"

## Questions

* ~~Is relying on `getGroupingBlockName()` from the package [here](https://github.com/WordPress/gutenberg/pull/16278/files#diff-2be030b99c851ec84dc58d21ae932a93R58) appropriate? I'm concerned I'm not using a `select()` here.~~ Now uses `select()`.
* Should we have some kind of validation in place before we allow custom blocks to be set as the Grouping Block? At the moment you can register any old Block but only Blocks which support the necessary transforms and utilise `InnerBlocks` are allowed. For example, you could _try_ and register `coblocks/column` to act as the Grouping Block (it supports `InnerBlocks` after all) but within the correct `convert()` (transform) definition it won't function as you'd expect. 
* Where should I write documentation to cover this feature (if at all)?


## Types of changes

New feature (non-breaking change which adds functionality).

## Checklist:
